python_modules/dagster/dagster/_core/asset_graph_view/asset_graph_view.py

```diff
@@ -0,0 +1,297 @@
+    respect to the effective time.
+
+    Last event id: Our event log has a monotonically increasing event id. This is used to
+    cursor the event log. This event_id is also propogated to derived tables to indicate
+    when that record is valid.  This allows us to query the state of the event log
+    at a given point in time.
+
+    Note that insertion time of the last_event_id is not the same as the effective time.
+
+    A last_event_id of None indicates that the reads will be volatile will immediately
+    reflect any subsequent writes.
+    """
+
+    effective_dt: datetime
+    last_event_id: Optional[int]
+
+
+# We reserve the right to constraints on the AssetSubset that we are going to use
+# in AssetSlice internals. Adding a NewType enforces that we do that conversion
+# in one spot (see _slice_from_subset)
+_AssetSliceCompatibleSubset = NewType("_AssetSliceCompatibleSubset", ValidAssetSubset)
+
+
+def _slice_from_subset(asset_graph_view: "AssetGraphView", subset: AssetSubset) -> "AssetSlice":
+    valid_subset = subset.as_valid(
+        asset_graph_view.asset_graph.get(subset.asset_key).partitions_def
+    )
+    return AssetSlice(asset_graph_view, _AssetSliceCompatibleSubset(valid_subset))
+
+
+class AssetSlice:
+    """An asset slice represents a set of partitions for a given asset key. It is
+    tied to a particular instance of an AssetGraphView, and is read-only.
+
+    With an asset slice you are able to traverse the set of partitions resident
+    in an asset graph at any point in time.
+
+    ```python
+        asset_graph_view_t0 = AssetGraphView.for_test(defs, effective_dt=some_date())
+
+        some_asset_slice = asset_graph_view_to.get_asset_slice(some_asset.key)
+
+        for parent_slice in some_asset_slice.compute_parent_slices().values():
+            # do something with the parent slice
+    ```
+
+    AssetSlice is read-only and tied to a specific AssetGraphView. Therefore
+    we can safely use cached methods and properties at will. However different methods
+    have different performance characterics so we have the following conventions:
+
+    Naming conventions
+    * Properties are guaranteed to be fast.
+    * Methods prefixed with `get_` do some work in-memory but are not hugely expensive.
+    * Methods prefixed with `compute_` do potentially expensive work, like compute
+      partition mappings and query the instance.
+
+      We also use this prefix to indicate that they fully materialize partition sets
+      These can potentially be very expensive if the underlying partition set has
+      an in-memory representation that involves large time windows. I.e. if the
+      underlying PartitionsSubset in the ValidAssetSubset is a
+      TimeWindowPartitionsSubset. Usage of these methods should be avoided if
+      possible if you are potentially dealing with slices with large time-based
+      partition windows.
+    """
+
+    def __init__(
```

Comment:
should we use a dataclass/namedtuple/basemodel here?

Reply:
They do not play nicely with @cached_method and @cached_property.

I actually experimented with declaring `__slots__` for the cached method instances, but that seemed like premature optimization.